tessel/src/protocol.rs

```diff
@@ -71,19 +107,63 @@ impl PortSocket {
         }
     }
 
-    pub fn write(&mut self, buffer: &[u8]) -> io::Result<()> {
-        try!(self.socket.write(buffer));
-        Ok(())
+    pub fn raw_write(&mut self, buffer: &[u8]) -> io::Result<()> {
+        self.socket.write_all(buffer)
     }
 
-    pub fn write_command(&mut self, cmd: command::Command, buffer: &[u8]) -> io::Result<()> {
-        try!(self.socket.write(&[cmd.0]));
-        try!(self.socket.write(buffer));
-        Ok(())
+    pub fn write_command(&mut self, cmd: Command) -> io::Result<()> {
+        let socket = &mut self.socket;
+        match cmd {
+            Nop => socket.write_all(&[raw_cmd::NOP]),
+            Flush => socket.write_all(&[raw_cmd::FLUSH]),
+            Rx(len) => socket.write_all(&[raw_cmd::RX, len]),
+            Echo(data) => {
+                assert!(data.len() <= u8::max_value() as usize);
+                try!(socket.write_all(&[raw_cmd::ECHO, data.len() as u8]));
+                socket.write_all(data)
+            },
+            Tx(data) => {
+                for slice in data.chunks(u8::max_value() as usize) {
+                    try!(socket.write_all(&[raw_cmd::TX, slice.len() as u8]));
+                    try!(socket.write_all(slice));
+                }
+                Ok(())
+            }
+            TxRx(data) => {
+                assert!(data.len() <= u8::max_value() as usize);
+                try!(socket.write_all(&[raw_cmd::TXRX, data.len() as u8]));
+                socket.write_all(data)
```

Comment:
Tx and TxRx are the only commands which require multiple calls to `write`. We could instead use a Vec, and build up the data we want to send first, then send in a batch, at the expense of having to do an allocation for the Vec (though its size could be determined, so no need for reallocation. We could also amortize the cost, by storing a buffer Vec in self.